engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,184 @@
+# PLAN-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zachary Burns</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0002-ethereum-tracing-cache.md">RFC-0002 Ethereum Tracing Cache</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/9">https://github.com/graphprotocol/rfcs/pull/9</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-20</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+
+## Summary
+
+Implements RFC-0002: Ethereum Tracing Cache
+
+## Implementation
+
+These changes happen within or near `ethereum_adapter.rs`, `store.rs` and `db_schema.rs`.
+
+### Limitations
+The problem of reorg turns out to be a particularly tricky one for the cache, mostly due to ranges of blocks being requested rather than individual hashes. To sidestep this problem, only blocks that are older than the reorg threshold will be eligible for caching.
+
+This workaround has an intended side effect. The target beneficiary of the workflow involving re-sync of recently deployed subgraphs. It may be undesirable to cache subgraphs which were not recently deployed and may never be re-synced. Such a subgraph would only incur the cost of caching, but never reap the benefits of reading from the cache. By only caching blocks outside the reorg threshold, only the subgraphs which have been recently deployed and are the most likely to benefit from caching are in fact cached.
+
+### EthereumTraceCache
+
+The implementation introduces the following trait, which is implemented primarily by `Store`.
+
+```rust
+use std::ops::Range;
```

Comment:
It will be cool to start using proper range types for block ranges. This should be `InclusiveRange` since block ranges are always inclusive.

Reply:
Good catch. I did not know that these were inclusive.

Reply:
I'm tempted to use Range in the API even if the Parity API uses InclusiveRange, if only because Range is so ubiquitous and well understood. It would be possible to convert at the boundary to adapt to the Parity API. My biggest concern is that would make migrating code a bit more error prone. Thoughts?